src/Stan/Analysis/Analyser.hs

```diff
@@ -286,129 +290,90 @@ declarations in a single pass.
+    matchOperator curNode
   where
     -- returns list of operator names defined in a single fixity declaration:
     -- infix 5 ***, +++, ???
-    matchInfix :: HieAST TypeIndex -> Slist OperatorDecl
-    matchInfix node@Node{..} = memptyIfFalse
+    matchInfix :: HieAST TypeIndex -> State VisitorState ()
+    matchInfix node@Node{..} = when
         (hieMatchPatternAst hie node fixity)
-        (S.concatMap nodeIds nodeChildren)
+        (traverse_ addFixity $ concatMap nodeIds nodeChildren)
 
     -- singleton or empty list with the top-level operator definition
-    matchOperator :: HieAST TypeIndex -> Slist OperatorDecl
-    matchOperator node@Node{..} = memptyIfFalse
+    matchOperator :: HieAST TypeIndex -> State VisitorState ()
+    matchOperator node@Node{..} = when
         (hieMatchPatternAst hie node typeSig)
-        (maybeToMonoid $ viaNonEmpty (extractOperatorName . head) nodeChildren)
-        -- first child of a parent is a name of a function/operator
+        (whenJust
+            -- do nothing when cannot extract name
+            -- first child of a parent is a name of a function/operator
+            (viaNonEmpty (extractOperatorName . head) nodeChildren)
+            -- add each operator decl from a list (should be singleton list)
+            (traverse_ (uncurry addOpDecl))
+        )
 
     -- return AST node identifier names as a sized list of texts
-    nodeIds :: HieAST TypeIndex -> Slist OperatorDecl
+    nodeIds :: HieAST TypeIndex -> [Text]
     nodeIds =
-        S.concatMap identifierName
+        concatMap fixityName
         . Map.keys
         . nodeIdentifiers
         . nodeInfo
 
-    identifierName :: Identifier -> Slist OperatorDecl
-    identifierName = \case
-        Left _ -> mempty
-        Right name -> S.one $ Fixity $ toText $ occNameString $ nameOccName name
+    fixityName :: Identifier -> [Text]
+    fixityName = \case
+        Left _ -> []
+        Right name -> [toText $ occNameString $ nameOccName name]
 
-    extractOperatorName :: HieAST TypeIndex -> Slist OperatorDecl
+    extractOperatorName :: HieAST TypeIndex -> [(Text, RealSrcSpan)]
     extractOperatorName Node{..} =
-        S.concatMap (topLevelOperatorName nodeSpan)
+        concatMap (topLevelOperatorName nodeSpan)
         $ Map.keys
         $ nodeIdentifiers nodeInfo
 
-    topLevelOperatorName :: RealSrcSpan -> Identifier -> Slist OperatorDecl
+    topLevelOperatorName :: RealSrcSpan -> Identifier -> [(Text, RealSrcSpan)]
     topLevelOperatorName srcSpan = \case
-        Left _ -> mempty
+        Left _ -> []
         Right name ->
             let occName = nameOccName name
-            in memptyIfFalse
-                (isSymOcc occName)  -- check if operator
-                (S.one $ Operator (toText $ occNameString occName) srcSpan)
-
--- | Either top-level operator or fixity declaration
-data OperatorDecl
-    = Fixity !Text
-    -- | Operator name with its position to display later
-    | Operator !Text !RealSrcSpan
```

Comment:
Nice that we can simplify this bit so much and get rid of some abstractions! :+1: